#991: Added Requirements Doc for Script Options parser (#469)
exasol/script-languages: exaudfclient/docs/script_options_requirments.md

# System Requirement Specification

This document outlines the user-centric requirements for the new Exasol UDF Client Script Options parser. It identifies the key features, high-level requirements, and user roles to ensure clarity and alignment with user needs.

## Roles

This section lists the roles that will interact with or benefit from the parser system.

### Database Administrator
Database Administrators manage the database environment and ensure the efficient execution of UDFs within Exasol, configuring and overseeing script execution.

### Data Scientist
Data Scientists develop and deploy UDFs in languages such as Java, Python, or R to process and analyze data within Exasol.

## Features

This section lists the key features of the new UDF Client Script Options parser which you would highlight in a product leaflet.

### General Script Options Parsing
`feat~general-script-options-parsing~1`

> 

Reviewer: The Developer can add additional Options in an easy and consistent way.

Needs: req

### Java-specific Script Options
`feat~java-specific-script-options~1`

> 

Reviewer: we need to see what goes here

Needs: req

## High-level Requirements

This section details the high-level requirements for the new parser system, linked to the features listed above.

### General Script Options Parsing
`req~general-script-options-parsing~1`

> The parser must correctly identify and handle Script Options with the syntax `%optionKey optionValue;`. It must also manage whitespace and ignore non-Script Options as specified.

Reviewer: syntax is incomplete, whitespaces between key and value

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Multiple and Duplicate Options Management
`req~multiple-duplicate-options-management~1`

> The parser must collect multiple Script Options with the same key and handle duplicates according to specific rules for different options.

Reviewer: Multiple requirements
- The parser must collect multiple Script Options with the same key. Note: the specific handling depends on the option handler
- The parser must collect multiple Script Options with the same key and Value. Note: the specific handling depends on the option handler

Reviewer: and the covers only `feat~general-script-options-parsing~1`

Needs: dsn

Covers:

> - `feat~general-script-options-parsing~1`

Reviewer: also only covers `feat~general-script-options-parsing~1`

- `feat~java-specific-script-options~1`

### Script Option Removal
`req~script-option-removal~1`

The parser must remove found Script Options from the original script code.

Needs: dsn

Covers:

> - `feat~general-script-options-parsing~1`

Reviewer: also only covers `feat~general-script-options-parsing~1`

- `feat~java-specific-script-options~1`

### Java %scriptclass Option Handling
`req~java-scriptclass-option-handling~1`

> The parser must correctly identify a single %scriptclass option and remove any additional occurrences of this option within the script code.

Reviewer: which one

Author: Changed to 
```suggestion
The parser must correctly identify the first %scriptclass option and remove any additional occurrences of this option within the script code.
```

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling
`req~java-jar-option-handling~1`

> The parser must find multiple %jar options, handle duplicates properly, and ensure the values match the Java CLASSPATH environment variable syntax.

Reviewer: what is with the order

Author: Order is defined in different requirements: `req~java-jar-option-handling-v1~1` and `req~java-jar-option-handling-v2~1`

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %jvmoption Handling
`req~java-jvmoption-handling~1`

The parser must find multiple %jvmoption options, allowing duplicates and maintaining order.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Handling
`req~java-import-option-handling~1`

> For each found %import option, the parser must request and replace the found Script Option with the referenced script code, also handling nested options appropriately.

Reviewer: split into two

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### New Parser Implementation
`req~new-parser-implementation~1`

> The new parser must be implemented using an existing, open-source parser that supports definition of Lexer and Parser Rules in C++ code without additional runtime dependencies.

Reviewer: add reason

Author: Done

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### New Parser Integration
`req~new-parser-integration~1`

> Ensure that the new parser integrates seamlessly into the Exasol UDF Client environment, meeting all specified dependency and embedding requirements.

Author: The AI didn't tell me :)
I will change this one to `req~linker-namespace-compatibility~1`

# System Requirement Specification

This document outlines the user-centric requirements for the new Exasol UDF Client Script Options parser. It identifies the key features, high-level requirements, and user roles to ensure clarity and alignment with user needs.

## Roles

This section lists the roles that will interact with or benefit from the parser system.

### Database Administrator
Database Administrators manage the database environment and ensure the efficient execution of UDFs within Exasol, configuring and overseeing script execution.

### Data Scientist
Data Scientists develop and deploy UDFs in languages such as Java, Python, or R to process and analyze data within Exasol.

## Features

This section lists the key features of the new UDF Client Script Options parser which you would highlight in a product leaflet.

### General Script Options Parsing
`feat~general-script-options-parsing~1`

> Script Options must be parsed according to syntax definition.

Author: Done

Developers can add additional Options in an easy and consistent way.

Needs: req

### Java-specific Script Options
`feat~java-specific-script-options~1`

> The parser must process all Java specific options correctly.

Author: Done

Needs: req

## High-level Requirements

This section details the high-level requirements for the new parser system, linked to the features listed above.

### General Script Options Parsing
`req~general-script-options-parsing~1`

The parser must correctly identify and handle Script Options with the syntax `%<optionKey><white spaces><optionValue>;`.

> The separator between

Reviewer: Incomplete

Author: Fixed

> The separator between

Author: Done

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### White Spaces
`req~white-spaces~1`

> The following is the list of white spaces:

Author: Done

|======================================================= 

> | Name         | C/Python/Java | ASCII Dec | ASCII Hex | 

Reviewer: C/Python/Java is weird

Author: Changed to `C syntax`

| tabulator    | '\t'          | 9         | 0x09      |
| vertical tab | '\v'          | 11         | 0x0b     |
| form feed    | '\f'          | 12         | 0x0c     |
| space        | ' '           | 30        | 0x20      |     
|======================================================= 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Leading White Spaces Options Parsing
`req~leading-white-spaces-script-options-parsing~1`

> The parser must recognize Script Options for lines starting with white space characters. 

Reviewer: ```suggestion
The parser must recognize Script Options for lines starting with white space characters before the Script Options. 
```

Author: Done

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Ignore anything which is not a Script Option
`req~ignore-none-script-options~1`

If there is any character in front of a Script Option which is not a white space, the parser must ignore the option(s). 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Multiple Line Script Options Parsing
`req~multiple-lines-script-options-parsing~1`

> The parser must recognize Script Options at any line in the given script code.

Author: Exactly. This is in order to be backwards compatible with the existing parser, and avoid major breakage of customer UDF's.

> The parser must recognize Script Options at any line in the given script code.

Reviewer: seee imports

> The parser must recognize Script Options at any line in the given script code.

Reviewer: Does a Note about the consequence makes sense. Plus a rational in the sense, option could get replaced with something for the final code

> The parser must recognize Script Options at any line in the given script code.

Author: Added a comment

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### White Spaces Options Parsing V1
`req~white-spaces-script-options-parsing-v1~1`

All white spaces between the option key and option value are to be interpreted as separator. 
White spaces between the option value and the terminating ";" are to be removed from the option value.

Needs: dsn

Tags: V1

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### White Spaces Options Parsing V2
`req~white-spaces-script-options-parsing-v2~1`

All white spaces between the option key and option value are to be ignored. The following rules for escape sequences at **the start** of a script optionValue are to be applied:
- '\ ' => space character
- '\t' => <tab> character
- '\f' => <form feed> character
- '\v' => <vertical tab> character

> White spaces between the option value and the terminating ";" shall remain.

Author: Done

Needs: dsn

Tags: V2

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

> ### Multiple and Options

Reviewer: ```suggestion
### Multiple Options
```

Author: Fixed

`req~multiple-options-management~1`

> The parser must collect multiple Script Options with the same key. Note: the specific handling depends on the option handler.

Author: Done

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Duplicate Options Management
`req~multiple-options-management~1`

> The parser must collect multiple Script Options with the same key and value. Note: the specific handling depends on the option handler.

Reviewer: see jvm options or jars

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Script Option Removal
`req~script-option-removal~1`

> The parser handler must remove found known Script Options from the original script code. Successfully parsed Script Options, which are not recognized by the parser handler shall remain in the script code.

Author: Note that I have used here the term `parser handler` as it is not the parser itself which removes the found options from the source script code, but the client.

Reviewer: This is not how the current parser works , if we want to change that we need to split the requirement per version and need to change the new implementation

Author: I extracted the second part:
```
Successfully parsed Script Options, which are not recognized by the parser handler shall remain in the script code.
```
into two new requirements: One for v1 (which behaves exactly as described here), and a new one for V2 which will throw an exception if an unknown option is detected (This is actually already implemented in that way in the new parser)

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`
- `feat~java-specific-script-options~1`

### Escape Sequence Script Options Parsing
`req~escape-sequence-script-options-parsing~1`

The following rules for escape sequences at any place within a script optionValue are to be applied:

> - '\n' => <line feed> character

Reviewer: I recognized something. What happens if user wants us \n inside a value without that it gets replaced

Author: I know, when writing the requirements yesterday I asked myself the same question. I need to fix that.
Need to add a lexer rule for `\\`

- '\r' => <carriage return> character
- '\;' => ';' character

Needs: dsn

Tags: V2

Covers:
- `feat~general-script-options-parsing~1`

### Java %scriptclass Option Handling V1
`req~java-scriptclass-option-handling-v1~1`

The Java parser handler must correctly identify the first `%scriptclass` option and remove only this single instance from the script code. Any further occurrences of `%scriptclass` option shall stay in the source script code.

Needs: dsn

Tags: V1

Covers:
- `feat~java-specific-script-options~1`

### Java %scriptclass Option Handling V2
`req~java-scriptclass-option-handling-v2~1`

The Java parser handler must correctly identify the first `%scriptclass` option and remove any additional occurrences of this option within the script code.

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling V1
`req~java-jar-option-handling-v1~1`

> The Java parser handler must find multiple %jar options. The values are to be interpreted as the Java CLASSPATH: `<file1>:<file2>:...:<filen>`.

Author: Done

> The Java parser handler shall split the entries by the colon character. The Java parser handler shall identify duplicated files and order the result of all `%jar` options alphabetically.

Author: Done

Needs: dsn

Tags: V1

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling V2
`req~java-jar-option-handling-v2~1`

The Java parser handler must find multiple %jar options. The values are to be interpreted as the Java CLASSPATH: `<file1>:<file2>:...:<filen>`.
The Java parser handler shall split the entries by the colon character. The Java parser handler must keep duplicates. The order of the entries must not change.

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Trailing White Space Handling
`req~java-jar-option-trailing-white-space-handling~1`

The Java parser handler must remove trailing white spaces for `%jar` option values if they are part of the escape sequence '\ '. Escape sequences at the end of a found `%jar` option of the form `\ ` must be replaced with ' '.

> This approach provides backwards compatibility for most existing UDF's from customers.

Author: Now that I have written down the requirement, I ask myself: Why do we do this only for the `%jar` option, and not for `%scriptclass`, `%jvmoption` and `%import`?

Author: Added more details about format for `%scriptclass` and `import`. `jvmoption` will be handled later

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

Depends: 
- `req~white-spaces-script-options-parsing-v2~1`
### Java %jvmoption Handling
`req~java-jvmoption-handling~1`

> The Java parser handler must find multiple %jvmoption options, allowing duplicates and maintaining order.

Author: Done

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Handling
`req~java-import-option-replace-referenced-scripts~1`

For each found %import option, the Java parser handler must request and replace the referenced scripts recursively. This means,
if the referenced scripts contain also `%import` options, the implementation must replace those, too.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Handling
`req~java-import-option-handling~1`

> For each found %import option, the Java parser handler must handle nested Script Options appropriately:

Author: Done

1. `%scriptclass` option must be ignored, but removed from the script code.
2. All other options must be handled as if they were part of the source script.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Existing Parser Library Dependencies
`req~existing-parser-library-dependencies~1`

> The new parser must be implemented using an existing, open-source parser that supports definition of Lexer and Parser Rules in C++ code without additional runtime dependencies.

Author: @tkilias , what is your opinion about that?

> The new parser must be implemented using an existing, open-source parser that supports definition of Lexer and Parser Rules in C++ code without additional runtime dependencies.

Reviewer: The requirement is that the parser needs to be usable in open source and closed source projects.

> The new parser must be implemented using an existing, open-source parser that supports definition of Lexer and Parser Rules in C++ code without additional runtime dependencies.

Author: Changed the requirment to simply:
```
### Existing Parser Library License
`req~existing-parser-library-license~1`

The parser needs to be usable in open source and closed source projects.

Needs: dsn

Tags: V2

Covers:
- `feat~general-script-options-parsing~1`
```

Needs: dsn

Tags: V1

Covers:
- `feat~java-specific-script-options~1`

### Java %scriptclass Option Handling V2
`req~java-scriptclass-option-handling-v2~1`

The Java parser handler must correctly identify the first `%scriptclass` option and remove any additional occurrences of this option within the script code.
The value should be handled according to the [Java specification for identifies](https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8).

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling V1
`req~java-jar-option-handling-v1~1`

The Java parser handler must find multiple %jar options. The values are to be interpreted as the Java CLASSPATH: `<file1>:<file2>:...:<filen>`.
The Java parser handler shall split the entries by the colon character. The Java parser handler shall identify duplicated files and order the result of all `%jar` options alphabetically.

Needs: dsn

Tags: V1

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling V2
`req~java-jar-option-handling-v2~1`

> 

Reviewer: I have the feeling both %jar option handling should be split and the first part can be common requirements for both

Author: Done

The Java parser handler must find multiple %jar options. The values are to be interpreted as the Java CLASSPATH: `<file1>:<file2>:...:<filen>`.
The Java parser handler shall split the entries by the colon character. The Java parser handler must keep duplicates. The order of the entries must not change.

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Trailing White Space Handling
`req~java-jar-option-trailing-white-space-handling~1`

> The Java parser handler must remove trailing white spaces for `%jar` option values if they are part of the escape sequence '\ '. Escape sequences at the end of a found `%jar` option of the form `\ ` must be replaced with ' '.

Reviewer: Are you sure with trimming the spaces at the.beginning. a relative path could still have spaces, however, how would a jar land at that path in the udf

Author: The requirement is a bit tricky, I haven't mentioned whitespaces at the beginning because this is already covered by `req~white-spaces-script-options-parsing-v2~1`. With the legacy implementation trimming always happened. Now you can explicitely define whitespace at the beginning of an option by using the escape sequences.

This approach provides backwards compatibility for most existing UDF's from customers.

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

Depends: 
- `req~white-spaces-script-options-parsing-v2~1`

### Java %jvmoption Handling
`req~java-jvmoption-handling~1`

The Java parser handler must find multiple %jvmoption options, allowing duplicates and maintaining order.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Replace Referenced Sripts
`req~java-import-option-replace-referenced-scripts~1`

For each found %import option, the Java parser handler must request and replace the referenced scripts recursively. This means,

> if the referenced scripts contain also `%import` options, the implementation must replace those, too.

Reviewer: Ignore already imported scripts is missing as a requirement

Author: Done. Added in `req~java-import-option-handling~1`

> The referenced script name should be handled according to the [Exasol SQL identifier specification](https://docs.exasol.com/db/latest/sql_references/basiclanguageelements.htm#SQLidentifier).

Reviewer: Separate requirement

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Handling
`req~java-import-option-handling~1`

For each found %import option, the Java parser handler must handle nested Script Options appropriately:
1. `%scriptclass` option must be ignored, but removed from the script code.
2. All other options must be handled as if they were part of the source script.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Existing Parser Library Dependencies
`req~existing-parser-library-dependencies~1`

The new parser must be implemented using an existing, open-source parser that supports definition of Lexer and Parser Rules in C++ code without additional runtime dependencies.
The implementation needs to be Open Source because the projects where the parser will be used are mainly Open Source, too. 

> It is important to avoid additional runtime dependencies, as this would complicate the setup and maintenance of the runtime environment of the UDF client (aka Script Languages Container).

Reviewer: We are also limited to certain licenses that can also used with closed source. So, a permissive license without copy left.

Author: Done

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Script Option Removal
`req~script-option-removal~1`

The parser handler must remove found known Script Options from the original script code. 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`
- `feat~java-specific-script-options~1`

### Script Option Unknown Options Behavior V1
`req~script-option-unknown-options-behvaior-v1~1`

Unknown script options must remain untouched in the script code. The Java compiler is supposed to throw an error message during the compilation. 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`
- `feat~java-specific-script-options~1`

Tags: V1

### Script Option Unknown Options Behavior V2
`req~script-option-unknown-options-behvaior-v2~1`

The parser handler must detect unknown script options and throw an exception if such an options is found. 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`
- `feat~java-specific-script-options~1`

Tags: V2

### Escape Sequence Script Options Parsing
`req~escape-sequence-script-options-parsing~1`

> The following rules for escape sequences at any place within a script optionValue are to be applied:

Reviewer: ```suggestion
The following rules for escape sequences at any place within a script option value are to be applied:
```

Author: Fixed

### Java %jar Option Handling Multiple Options
`req~java-jar-option-handling-multiple-options~1`

The Java parser handler must find multiple `%jar` options. The values are to be interpreted as the Java CLASSPATH: `<file1>:<file2>:...:<filen>`. The Java parser handler shall split the entries by the colon character.
Compare [OpenJdk implementation](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/jdk/internal/loader/URLClassPath.java#L174) of parsing the classpath.

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling V1
`req~java-jar-option-handling-v1~1`

The Java parser handler shall unify duplicated files and order the result of all `%jar` options alphabetically.

Needs: dsn

Tags: V1

Covers:
- `feat~java-specific-script-options~1`

Depends:
- `req~java-jar-option-handling-multiple-options~1`

### Java %jar Option Handling V2
`req~java-jar-option-handling-v2~1`

The Java parser handler must keep duplicates. The order of the entries must not change.

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

Depends:
- `req~java-jar-option-handling-multiple-options~1`

> 

Reviewer: Jar option in v2 needs to be able to deal with whitespace in the jar paths. This requirement is missing.

### Java %jvmoption Handling
`req~java-jvmoption-handling~1`

The Java parser handler must find multiple `%jvmoption` options, allowing duplicates and maintaining order.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Replace Referenced Sripts
`req~java-import-option-replace-referenced-scripts~1`

For each found `%import` option, the Java parser handler must request and replace the referenced scripts recursively. This means, if the referenced scripts contain also `%import` options, the implementation must replace those, too.

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Referenced Sripts Name
`req~java-import-option-referenced-scripts-name~1`

The referenced script name should be handled according to the [Exasol SQL identifier specification](https://docs.exasol.com/db/latest/sql_references/basiclanguageelements.htm#SQLidentifier).

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Java %import Option Handling
`req~java-import-option-handling~1`

For each found `%import` option, the Java parser handler must handle nested Script Options appropriately:

> 1. `%scriptclass` option must be ignored, but removed from the script code.

Reviewer: Was this also in v1 the case?

Author: Splitted it up

2. All other options must be handled as if they were part of the source script.
3. Already imported scripts must not be imported again, but the `%import` statement must be removed

Needs: dsn

Covers:
- `feat~java-specific-script-options~1`

### Existing Parser Library License
`req~existing-parser-library-license~1`

The parser needs to be usable in open source and closed source projects.

Needs: dsn

Tags: V2

Covers:
- `feat~general-script-options-parsing~1`

> ### Existing Parser Linker Namespace Compatibility

Reviewer: Technical constraint, so this goes to the design

|======================================================= 
| Name         | C syntax      | ASCII Dec | ASCII Hex | 
| tabulator    | '\t'          | 9         | 0x09      |
| vertical tab | '\v'          | 11         | 0x0b     |
| form feed    | '\f'          | 12         | 0x0c     |
| space        | ' '           | 30        | 0x20      |     
|======================================================= 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Leading White Spaces Options Parsing
`req~leading-white-spaces-script-options-parsing~1`

The parser must accept Script Options for lines starting with any number of white space characters before the Script Options.

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Ignore anything which is not a Script Option
`req~ignore-none-script-options~1`

If there is any character in front of a Script Option which is not a white space, the parser must ignore the option(s). 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Multiple Line Script Options Parsing
`req~multiple-lines-script-options-parsing~1`

> The parser must recognize Script Options at any line in the given script code. This is especially important because the `%import` option (see requirement `req~java-import-option-replace-referenced-scripts~1`) might replace options with Java code in the final script code.

Reviewer: ```suggestion
The parser must recognize Script Options at any line in the given script code. 

Rationale: This is especially important because the `%import` option (see requirement `req~java-import-option-replace-referenced-scripts~1`) might replace options with Java code in the final script code.
```

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### White Spaces Options Parsing V1
`req~white-spaces-script-options-parsing-v1~1`

All white spaces between the option key and option value are to be interpreted as separator. 
White spaces between the option value and the terminating ";" are to be removed from the option value.

Needs: dsn

Tags: V1

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### White Spaces Options Parsing V2
`req~white-spaces-script-options-parsing-v2~1`

> All white spaces between the option key and option value are to be ignored. The following rules for escape sequences at **the start** of a script optionValue are to be applied:

Reviewer: ```suggestion
All white spaces between the option key and option value are to be ignored. The following rules for escape sequences at **the start** of a script option value are to be applied:
```

- '\ ' => space character
- '\t' => <tab> character
- '\f' => <form feed> character
- '\v' => <vertical tab> character

White spaces in the middle of the option value and between the option value and the terminating ";" shall be interpreted as part of the value.

> The rationale is that the new version of the parser should be as much as possible backwards compatible to V1, because it will simplify migration of existing UDF's.

Reviewer: ```suggestion

Rationale: The new version of the parser should be as much as possible backwards compatible to V1, because it will simplify migration of existing UDF's.
```

# System Requirement Specification

This document outlines the user-centric requirements for the new Exasol UDF Client Script Options parser. It identifies the key features, high-level requirements, and user roles to ensure clarity and alignment with user needs.

## Roles

This section lists the roles that will interact with or benefit from the parser system.

### Database Administrator
Database Administrators manage the database environment and ensure the efficient execution of UDFs within Exasol, configuring and overseeing script execution.

> ### Data Scientist

Reviewer: Developer role potentially missing

Author: Hm, I'am not sure if we need a Developer Role. The Data Scientist role already covers the development of UDF's by customers.

Reviewer: Developer is the person who integrates the lib into other software

Author: Added the role

### Java-specific Script Options
`feat~java-specific-script-options~1`

The parser must process all Java-specific options correctly.

Needs: req

## High-level Requirements

This section details the high-level requirements for the new parser system, linked to the features listed above.

### General Script Options Parsing
`req~general-script-options-parsing~1`

The parser must correctly identify and handle Script Options with the syntax `%<optionKey><white spaces><optionValue>;`.

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### White Spaces
`req~white-spaces~1`

The parser must treat the following list of white spaces as token separator:
|======================================================= 
| Name         | C syntax      | ASCII Dec | ASCII Hex | 
| tabulator    | '\t'          | 9         | 0x09      |
| vertical tab | '\v'          | 11         | 0x0b     |
| form feed    | '\f'          | 12         | 0x0c     |
| space        | ' '           | 30        | 0x20      |     
|======================================================= 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### White Spaces Separator
`req~white-spaces-separator~1`

> A Script Options can have one or more white spaces between the Script Option Key `<optionKey>` and Script Option Value `<<optionValue>`.

Reviewer: ```suggestion
A Script Options can have one or more white spaces between the Script Option Key `<optionKey>` and Script Option Value `<optionValue>`.
```

Author: Fixed

The parser must recognize those white spaces as separator(s).

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Leading White Spaces Options Parsing
`req~leading-white-spaces-script-options-parsing~1`

The parser must accept Script Options for lines starting with any number of white space characters before the Script Options.

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Ignore anything which is not a Script Option
`req~ignore-none-script-options~1`

If there is any character in front of a Script Option which is not a white space, the parser must ignore the option(s). 

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Multiple Line Script Options Parsing
`req~multiple-lines-script-options-parsing~1`

The parser must recognize Script Options at any line in the given script code.

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### White Spaces Options Parsing V1
`req~white-spaces-script-options-parsing-v1~1`

> All white spaces between the option key and option value are to be interpreted as separator. 

Reviewer: Somewhat of a duplicate of `req~white-spaces-separator~1`

White spaces between the option value and the terminating ";" are to be removed from the option value.

Needs: dsn

Tags: V1

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### White Spaces Options Parsing V2
`req~white-spaces-script-options-parsing-v2~1`

> All white spaces between the option key and option value are to be ignored. The following rules for escape sequences at **the start** of a script optionValue are to be applied:

Reviewer: Its actually not only at the start I guess, but we need it at the start

Author: Escape sequences which are not at the start are not converted.
Compare https://github.com/exasol/script-languages/blob/f2ba8d2b531ecde327ca11f84f50c31594c6f80d/exaudfclient/base/script_options_parser/ctpg/script_option_lines_ctpg.cc#L146 versus https://github.com/exasol/script-languages/blob/f2ba8d2b531ecde327ca11f84f50c31594c6f80d/exaudfclient/base/script_options_parser/ctpg/script_option_lines_ctpg.cc#L140

Reviewer: add rationale, we wanted to stay as much as possible backwards compatible

Author: Done

- '\ ' => space character
- '\t' => <tab> character
- '\f' => <form feed> character
- '\v' => <vertical tab> character

> White spaces between the option value and the terminating ";" shall be interpreted as part of the value.

Reviewer: In-between of the value is missing

Author: Fixed

Needs: dsn

Tags: V2

Covers:
- `feat~general-script-options-parsing~1`

Depends:
 - `req~white-spaces~1`

### Multiple Options
`req~multiple-options-management~1`

The parser must collect multiple Script Options with the same key.

Comment: 
"Collecting" in this context means that a merging strategy must be applied. We don't want multiple options with the same key to result in simply overwriting or discarding values. Please note that the specific handling depends on the individual option handler.

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Duplicate Options Management
`req~multiple-options-management~1`

The parser must collect multiple Script Options with the same key and same/different value. Note: the specific handling depends on the option handler.

Needs: dsn

Covers:
- `feat~general-script-options-parsing~1`

### Script Option Removal
`req~script-option-removal~1`

> The parser handler must remove found known Script Options from the original script code. 

Reviewer: Hm, this implemented by the handler, do we have a reusable implementation for that, otherwise we need to reimplement this for each handler

Author: Not yet, but should be an easy refactoring to do so.

Needs: dsn

Tags: V2

Covers:
- `feat~general-script-options-parsing~1`

### Java %scriptclass Option Handling V1
`req~java-scriptclass-option-handling-v1~1`

The Java parser handler must correctly identify the first `%scriptclass` option and remove only this single instance from the script code. Any further occurrences of `%scriptclass` option shall stay in the source script code.
The value should be handled according to the [Java specification for identifies](https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8).

Needs: dsn

Tags: V1

Covers:
- `feat~java-specific-script-options~1`

### Java %scriptclass Option Handling V2
`req~java-scriptclass-option-handling-v2~1`

The Java parser handler must correctly identify the first `%scriptclass` option and remove any additional occurrences of this option within the script code.
The value should be handled according to the [Java specification for identifies](https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8).

Needs: dsn

Tags: V2

Covers:
- `feat~java-specific-script-options~1`

### Java %jar Option Handling Multiple Options
`req~java-jar-option-handling-multiple-options~1`

> The Java parser handler must find multiple `%jar` options. The values are to be interpreted as the Java CLASSPATH: `<file1>:<file2>:...:<filen>`. The Java parser handler shall split the entries by the colon character.

Reviewer: Maybe add note to source code which defines how colon is handled in the jvm

Author: Done
